packages/plugin-ext/src/main/browser/editor/untitled-resource.ts

```diff
@@ -14,39 +14,116 @@
-import { injectable } from 'inversify';
 import { Schemes } from '../../../common/uri-components';
+import { FileResource, FileResourceResolver } from '@theia/filesystem/lib/browser';
 
-const resources = new Map<string, UntitledResource>();
 let index = 0;
+
 @injectable()
 export class UntitledResourceResolver implements ResourceResolver {
-    resolve(uri: URI): Resource | Promise<Resource> {
-        if (uri.scheme === Schemes.UNTITLED) {
-            return resources.get(uri.toString())!;
+
+    @inject(FileResourceResolver)
+    protected readonly fileResourceResolver: FileResourceResolver;
+
+    protected readonly resources = new Map<string, UntitledResource>();
+
+    async resolve(uri: URI): Promise<UntitledResource> {
+        if (uri.scheme !== Schemes.UNTITLED) {
+            throw new Error('The given uri is not untitled file uri: ' + uri);
+        } else {
+            const untitledResource = this.resources.get(uri.toString());
+            if (!untitledResource) {
+                return this.createUntitledResource(this.fileResourceResolver, '', '', uri);
+            } else {
+                return untitledResource;
+            }
         }
-        throw new Error(`scheme ${uri.scheme} is not '${Schemes.UNTITLED}'`);
+    }
+
+    async  createUntitledResource(fileResourceResolver: FileResourceResolver, content?: string, language?: string, uri?: URI): Promise<UntitledResource> {
+        let extension;
+        if (language) {
+            for (const lang of monaco.languages.getLanguages()) {
+                if (lang.id === language) {
+                    if (lang.extensions) {
+                        extension = lang.extensions[0];
+                        break;
+                    }
+                }
+            }
+        }
+        return new UntitledResource(this.resources, uri ? uri : new URI().withScheme(Schemes.UNTITLED).withPath(`/Untitled-${index++}${extension ? extension : ''}`),
+            fileResourceResolver, content);
     }
 }
 
 export class UntitledResource implements Resource {
 
-    constructor(public uri: URI, private content?: string) {
-        resources.set(this.uri.toString(), this);
-    }
+    private fileResource?: FileResource;
+
+    protected readonly onDidChangeContentsEmitter = new Emitter<void>();
+    readonly onDidChangeContents: Event<void> = this.onDidChangeContentsEmitter.event;
 
-    readContents(options?: { encoding?: string | undefined; } | undefined): Promise<string> {
-        return Promise.resolve(this.content ? this.content : '');
+    constructor(private resources: Map<string, UntitledResource>, public uri: URI, private fileResourceResolver: FileResourceResolver, private content?: string) {
+        this.resources.set(this.uri.toString(), this);
     }
 
     dispose(): void {
-        resources.delete(this.uri.toString());
+        this.resources.delete(this.uri.toString());
+        this.onDidChangeContentsEmitter.dispose();
```

Comment:
`fileResource` should be disposed as well, otherwise it will dangle after editor is closed and fire events

Reply:
done